File: Examples/Advanced/Run-Cmd-And-Export-Log/Python/run_cmd_and_export_log.py
#!/usr/bin/env python3
"""Run device commands through netLD/ThirdEye and export the returned logs."""

from __future__ import annotations

import copy
import json
import os
import re
import sys
import time
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

try:
    import requests
except ModuleNotFoundError:
    requests = None  # type: ignore[assignment]

try:
    from dotenv import load_dotenv
except ModuleNotFoundError:
    load_dotenv = None  # type: ignore[assignment]


class ExampleError(RuntimeError):
    pass


def env_bool(name: str, default: bool = False) -> bool:
    value = os.getenv(name)
    if value is None or not value.strip():
        return default
    return value.strip().lower() in {"1", "true", "yes", "on"}


def safe_filename(value: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9._-]+", "_", value.strip()).strip("._")
    return cleaned or "device"


def build_job(network: str, target: str, commands: list[str], backup: bool = False) -> dict[str, Any]:
    if not commands:
        raise ExampleError("The command file contains no commands.")
    return {
        "jobName": f"API Commands - {target}",
        "managedNetworks": [network],
        "jobType": "Script Tool Job",
        "description": "Ad hoc command execution from the NetLD SDK advanced example",
        "jobParameters": {
            "tool": "org.ziptie.tools.scripts.commandRunner",
            "managedNetwork": network,
            "ipResolutionScheme": "ipCsv",
            "ipResolutionData": f'"{target}@{network}"',
            "backupOnCompletion": str(backup).lower(),
            "input.commandList": "\n".join(commands),
        },
    }


@dataclass(frozen=True)
class Config:
    base_url: str
    api_key: str
    network: str
    target: str
    commands: list[str]
    output_dir: Path
    run_job: bool
    poll_seconds: float
    timeout_seconds: float
    backup: bool

    @classmethod
    def from_env(cls, env_path: Path) -> "Config":
        if load_dotenv is None:
            raise ExampleError("Install the Python dependencies from requirements.txt.")
        load_dotenv(env_path, override=True)
        required = {}
        for name in ("NETLD_BASE_URL", "NETLD_API_KEY", "NETLD_TARGET"):
            value = os.getenv(name, "").strip()
            if not value:
                raise ExampleError(f"Set {name} in .env before running this example.")
            required[name] = value

        command_path = Path(os.getenv("NETLD_COMMAND_FILE", "commands.txt"))
        output_dir = Path(os.getenv("NETLD_OUTPUT_DIR", "output"))
        if not command_path.is_absolute():
            command_path = env_path.parent / command_path
        if not output_dir.is_absolute():
            output_dir = env_path.parent / output_dir
        try:
            commands = [line.strip() for line in command_path.read_text(encoding="utf-8").splitlines() if line.strip()]
        except OSError as exc:
            raise ExampleError(f"Could not read the command file: {command_path}") from exc

        return cls(
            base_url=required["NETLD_BASE_URL"].rstrip("/"),
            api_key=required["NETLD_API_KEY"],
            network=os.getenv("NETLD_NETWORK", "Default").strip() or "Default",
            target=required["NETLD_TARGET"],
            commands=commands,
            output_dir=output_dir,
            run_job=env_bool("NETLD_RUN_JOB"),
            poll_seconds=float(os.getenv("NETLD_POLL_SECONDS", "2")),
            timeout_seconds=float(os.getenv("NETLD_WAIT_TIMEOUT_SECONDS", "300")),
            backup=env_bool("NETLD_BACKUP_ON_COMPLETION"),
        )


class NetLDClient:
    def __init__(self, base_url: str, api_key: str, timeout: float = 30):
        if requests is None:
            raise ExampleError("Install the Python dependencies from requirements.txt.")
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Authorization": f"Bearer {api_key}"})

    def login(self) -> None:
        self._request("GET", "/rest")

    def _request(self, method: str, path: str, **kwargs: Any):
        try:
            response = self.session.request(
                method,
                f"{self.base_url}{path}",
                timeout=self.timeout,
                allow_redirects=False,
                **kwargs,
            )
        except requests.RequestException as exc:
            raise ExampleError(f"Could not reach {self.base_url}.") from exc
        if 300 <= response.status_code < 400:
            raise ExampleError(f"Request redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise ExampleError(f"Request failed with HTTP {response.status_code}.") from exc
        return response

    def call(self, method: str, **parameters: Any) -> Any:
        response = self._request(
            "POST",
            "/rest",
            headers={"Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "method": method, "params": parameters, "id": str(uuid.uuid4())},
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise ExampleError(f"{method} returned invalid JSON.") from exc
        if data.get("error"):
            raise ExampleError(f"{method} failed: {json.dumps(data['error'])}")
        return data.get("result")

    def download_detail(self, execution_id: int, record_id: int) -> str:
        response = self._request(
            "GET",
            "/servlet/pluginDetail",
            params={"executionId": execution_id, "recordId": record_id},
        )
        return response.content.decode("utf-8", errors="replace")


def wait_for_completion(client: NetLDClient, execution: dict[str, Any], poll: float, timeout: float):
    deadline = time.monotonic() + timeout
    current = copy.deepcopy(execution)
    while current.get("endTime") is None:
        if time.monotonic() >= deadline:
            raise ExampleError(f"Execution {execution.get('id')} did not finish within {timeout:g} seconds.")
        time.sleep(poll)
        current = client.call("Scheduler.getExecutionDataById", executionId=execution["id"])
        if not current:
            raise ExampleError(f"Scheduler returned no data for execution {execution['id']}.")
    return current


def export_details(client: NetLDClient, execution: dict[str, Any], output_dir: Path) -> list[Path]:
    details = client.call("Plugins.getExecutionDetails", executionId=execution["id"]) or []
    if not details:
        raise ExampleError(f"No device output was returned for execution {execution['id']}.")
    output_dir.mkdir(parents=True, exist_ok=True)
    paths = []
    for detail in details:
        timestamp = datetime.fromtimestamp(
            int(detail.get("startTime") or execution.get("startTime") or 0) / 1000,
            tz=timezone.utc,
        ).strftime("%Y%m%dT%H%M%SZ")
        identity = safe_filename(
            f"{detail.get('managedNetwork', 'network')}_{detail.get('ipAddress', 'device')}"
        )
        path = output_dir / f"{timestamp}_{execution['id']}_{detail['id']}_{identity}.log"
        content = client.download_detail(execution["id"], detail["id"])
        path.write_text(content, encoding="utf-8")
        paths.append(path)
    return paths


def main() -> int:
    try:
        config = Config.from_env(Path(__file__).with_name(".env"))
        job = build_job(config.network, config.target, config.commands, config.backup)
        print(json.dumps(job, indent=2))
        if not config.run_job:
            print("Dry run only. Set NETLD_RUN_JOB=true after reviewing this job.")
            return 0

        client = NetLDClient(config.base_url, config.api_key)
        client.login()
        execution = client.call("Scheduler.runNow", jobData=job)
        final = wait_for_completion(client, execution, config.poll_seconds, config.timeout_seconds)
        print(json.dumps(final, indent=2))
        for path in export_details(client, final, config.output_dir):
            print(f"Wrote {path}")
        if final.get("status") not in (None, "OK"):
            raise ExampleError(f"Execution completed with status {final.get('status')}.")
        return 0
    except (ExampleError, OSError, ValueError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: Examples/Advanced/Run-Cmd-And-Export-Log/nodeJS/run-cmd-and-export-log.mjs
#!/usr/bin/env node

import { randomUUID } from "node:crypto";
import { readFile, mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath, pathToFileURL } from "node:url";

export class ExampleError extends Error {}

export function safeFilename(value) {
  return value.trim().replace(/[^A-Za-z0-9._-]+/g, "_").replace(/^[._]+|[._]+$/g, "") || "device";
}

export function buildJob(network, target, commands, backup = false) {
  if (!commands.length) throw new ExampleError("The command file contains no commands.");
  return {
    jobName: `API Commands - ${target}`,
    managedNetworks: [network],
    jobType: "Script Tool Job",
    description: "Ad hoc command execution from the NetLD SDK advanced example",
    jobParameters: {
      tool: "org.ziptie.tools.scripts.commandRunner",
      managedNetwork: network,
      ipResolutionScheme: "ipCsv",
      ipResolutionData: `"${target}@${network}"`,
      backupOnCompletion: String(backup),
      "input.commandList": commands.join("\n"),
    },
  };
}

function envBoolean(name, defaultValue = false) {
  const value = process.env[name];
  return value == null || !value.trim()
    ? defaultValue
    : ["1", "true", "yes", "on"].includes(value.trim().toLowerCase());
}

async function loadDotEnv(file) {
  let text;
  try {
    text = await readFile(file, "utf8");
  } catch (error) {
    if (error.code === "ENOENT") return;
    throw error;
  }
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const offset = line.indexOf("=");
    process.env[line.slice(0, offset).trim()] = line
      .slice(offset + 1)
      .trim()
      .replace(/^(['"])(.*)\1$/, "$2");
  }
}

async function loadConfig(directory) {
  await loadDotEnv(path.join(directory, ".env"));
  for (const name of ["NETLD_BASE_URL", "NETLD_API_KEY", "NETLD_TARGET"]) {
    if (!process.env[name]?.trim()) {
      throw new ExampleError(`Set ${name} in .env before running this example.`);
    }
  }
  const commandFile = path.resolve(directory, process.env.NETLD_COMMAND_FILE || "commands.txt");
  const commands = (await readFile(commandFile, "utf8"))
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);
  return {
    baseUrl: process.env.NETLD_BASE_URL.replace(/\/$/, ""),
    apiKey: process.env.NETLD_API_KEY,
    network: process.env.NETLD_NETWORK?.trim() || "Default",
    target: process.env.NETLD_TARGET.trim(),
    commands,
    outputDir: path.resolve(directory, process.env.NETLD_OUTPUT_DIR || "output"),
    runJob: envBoolean("NETLD_RUN_JOB"),
    pollMilliseconds: Number(process.env.NETLD_POLL_SECONDS || 2) * 1000,
    timeoutMilliseconds: Number(process.env.NETLD_WAIT_TIMEOUT_SECONDS || 300) * 1000,
    backup: envBoolean("NETLD_BACKUP_ON_COMPLETION"),
  };
}

class NetLDClient {
  constructor(baseUrl, apiKey, timeout = 30000) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.timeout = timeout;
    this.cookies = "";
  }

  headers(extra = {}) {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      ...(this.cookies ? { Cookie: this.cookies } : {}),
      ...extra,
    };
  }

  async request(pathname, options = {}) {
    let response;
    try {
      response = await fetch(`${this.baseUrl}${pathname}`, {
        redirect: "manual",
        signal: AbortSignal.timeout(this.timeout),
        ...options,
        headers: this.headers(options.headers),
      });
    } catch {
      throw new ExampleError(`Could not reach ${this.baseUrl}.`);
    }
    if (response.status >= 300 && response.status < 400) {
      throw new ExampleError(`Request redirected to ${response.headers.get("location") || ""}.`);
    }
    if (!response.ok) throw new ExampleError(`Request failed with HTTP ${response.status}.`);
    const setCookies = response.headers.getSetCookie?.() || splitSetCookie(response.headers.get("set-cookie") || "");
    if (setCookies.length) this.cookies = setCookies.map((item) => item.split(";", 1)[0]).join("; ");
    return response;
  }

  async login() {
    await this.request("/rest");
  }

  async call(method, parameters = {}) {
    const response = await this.request("/rest", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ jsonrpc: "2.0", method, params: parameters, id: randomUUID() }),
    });
    const data = await response.json();
    if (data.error) throw new ExampleError(`${method} failed: ${JSON.stringify(data.error)}`);
    return data.result;
  }

  async downloadDetail(executionId, recordId) {
    const query = new URLSearchParams({ executionId, recordId });
    return (await this.request(`/servlet/pluginDetail?${query}`)).text();
  }
}

const sleep = (milliseconds) => new Promise((resolve) => setTimeout(resolve, milliseconds));

function splitSetCookie(header) {
  return header ? header.split(/,(?=\s*[^;=]+=[^;]+)/) : [];
}

async function waitForCompletion(client, execution, pollMilliseconds, timeoutMilliseconds) {
  const deadline = Date.now() + timeoutMilliseconds;
  let current = execution;
  while (current.endTime == null) {
    if (Date.now() >= deadline) {
      throw new ExampleError(`Execution ${execution.id} did not finish within ${timeoutMilliseconds / 1000} seconds.`);
    }
    await sleep(pollMilliseconds);
    current = await client.call("Scheduler.getExecutionDataById", { executionId: execution.id });
    if (!current) throw new ExampleError(`Scheduler returned no data for execution ${execution.id}.`);
  }
  return current;
}

async function exportDetails(client, execution, outputDir) {
  const details = (await client.call("Plugins.getExecutionDetails", { executionId: execution.id })) || [];
  if (!details.length) throw new ExampleError(`No device output was returned for execution ${execution.id}.`);
  await mkdir(outputDir, { recursive: true });
  const paths = [];
  for (const detail of details) {
    const date = new Date(Number(detail.startTime || execution.startTime || 0));
    const timestamp = date.toISOString().replace(/[-:]/g, "").replace(/\.\d{3}/, "");
    const identity = safeFilename(`${detail.managedNetwork || "network"}_${detail.ipAddress || "device"}`);
    const destination = path.join(outputDir, `${timestamp}_${execution.id}_${detail.id}_${identity}.log`);
    await writeFile(destination, await client.downloadDetail(execution.id, detail.id), "utf8");
    paths.push(destination);
  }
  return paths;
}

export async function main() {
  const directory = path.dirname(fileURLToPath(import.meta.url));
  const config = await loadConfig(directory);
  const job = buildJob(config.network, config.target, config.commands, config.backup);
  console.log(JSON.stringify(job, null, 2));
  if (!config.runJob) {
    console.log("Dry run only. Set NETLD_RUN_JOB=true after reviewing this job.");
    return;
  }
  const client = new NetLDClient(config.baseUrl, config.apiKey);
  await client.login();
  const execution = await client.call("Scheduler.runNow", { jobData: job });
  const final = await waitForCompletion(client, execution, config.pollMilliseconds, config.timeoutMilliseconds);
  console.log(JSON.stringify(final, null, 2));
  for (const destination of await exportDetails(client, final, config.outputDir)) {
    console.log(`Wrote ${destination}`);
  }
  if (final.status != null && final.status !== "OK") {
    throw new ExampleError(`Execution completed with status ${final.status}.`);
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch((error) => {
    console.error(`Error: ${error.message}`);
    process.exitCode = 1;
  });
}
